src/ContextProvider/OgContext.php

```diff
@@ -0,0 +1,209 @@
+   * context. We might then decide which group is the better candidate by
+   * looking at a URL query argument or inspecting the user's browsing history
+   * to see if they are coming from a group page.
+   *
+   * This discovery of groups is handled by OgGroupResolver plugins. Each plugin
+   * is responsible for discovering groups in a specific domain (e.g. find all
+   * groups belonging to the current route). The plugins that will be used for
+   * the discovery are configurable; they are listed under the 'group_resolvers'
+   * key in the 'og.settings' config. The plugins are ordered by priority,
+   * meaning that if two groups are relevant to the current context, then the
+   * plugin with the highest priority will decide which group is going to 'win'.
+   *
+   * Developers can customize the group context result by providing their own
+   * plugins and by activating, disabling or reordering the default ones.
+   *
+   * @return \Drupal\Core\Entity\EntityInterface|null
+   *   The group entity which is most relevant in the current context, or NULL
+   *   if no relevant group was found.
+   *
+   * @see \Drupal\og\OgGroupResolverInterface
+   */
+  protected function getBestCandidate() {
+    $collection = new OgResolvedGroupCollection();
+    $plugins = [];
+
+    // Retrieve the list of group resolvers. These are stored in config, and are
+    // ordered by priority.
+    $group_resolvers = $this->configFactory->get('og.settings')->get('group_resolvers');
+    $priority = 0;
+    foreach ($group_resolvers as $plugin_id) {
+      /** @var \Drupal\og\OgGroupResolverInterface $plugin */
+      if ($plugin = $this->pluginManager->createInstance($plugin_id)) {
+        $plugins[$plugin_id] = $plugin;
+
+        // Set the default vote weight according to the plugin's priority.
+        $collection->setVoteWeight($priority);
+
+        // Let the plugin do its magic.
+        $plugin->resolve($collection);
+
+        // If the plugin is certain that the candidate belongs to the current
+        // context, it can declare the search to be over.
+        if ($plugin->isPropagationStopped()) {
+          break;
+        }
+
+        // The next plugin we try will have a lower priority.
+        $priority--;
+      }
+    }
+
+    // @todo This should be moved to a plugin.
+    // Filter out inappropriate results.
+    array_filter($collection, function (array $candidate) {
+      // Filter out results that are not accessible by the current user.
+      return $candidate['entity']->access('view');
+    });
+
+    // @todo This should be handled in the collection.
+    // Since we have filtered the results by the current user, we need to add
+    // the user as a cache context.
+    $this->addCacheContextIds(['user']);
+
+    // @todo Move this in the collection? Seems to make sense, since the
+    //   collection has knowledge about the groups and votes.
+    // Find the best matching group by iterating over the candidates and return
```

Comment:
Yes, with a small difference, their vote will be depending on their position in the list, so the first plugin's vote will count for more than the second plugin's vote. This will allow us to easily choose the correct "winner" if we have two groups with the same number of votes.